File: leetcode-go/Q029-Divide-Two-Integers/Q029-Divide-Two-Integers.go
//package Q029_Divide_Two_Integers
package main

import "math"

func divide(dividend int, divisor int) int {
	if divisor == 0 || dividend == math.MinInt32 && divisor == -1 {
		return math.MaxInt32
	}

	sign := dividend > 0 && divisor > 0 || dividend < 0 && divisor < 0
	dvd, dvs := abs(dividend), abs(divisor)

	ans := 0
	for dvs <= dvd {
		tmp := dvs
		cnt := 1
		for dvd >= tmp<<1 {
			tmp <<= 1
			cnt <<= 1
		}
		dvd -= tmp
		ans += cnt
	}

	if sign {
		return ans
	}
	return 0 - ans
}

func abs(a int) int64 {
	if a > 0 {
		return int64(a)
	}
	return int64(0 - a)
}

File: src/facebook/Q029_Divide_Two_Integers.java
package facebook;

public class Q029_Divide_Two_Integers {

    public int divide(int dividend, int divisor) {
        if (divisor == 0 || dividend == Integer.MIN_VALUE && divisor == -1) {
            return Integer.MAX_VALUE;
        }
        int sign = (dividend > 0 && divisor > 0) || (dividend < 0 && divisor < 0) ? 1 : -1;
        long dvd = Math.abs((long) dividend);
        long dvs = Math.abs((long) divisor);
        int result = 0;
        while (dvd >= dvs) {
            long tmp = dvs, mul = 1;
            while (dvd >= (tmp << 1)) {
                tmp <<= 1;
                mul <<= 1;
            }
            dvd -= tmp;
            result += mul;
        }
        return result * sign;
    }

}
